test/matrix.lean

```diff
@@ -0,0 +1,169 @@
+/-
+manually ported from
+https://github.com/leanprover-community/mathlib/blob/4f4a1c875d0baa92ab5d92f3fb1bb258ad9f3e5b/test/matrix.lean
+-/
+import Mathlib.Data.Matrix.Notation
+-- import linear_algebra.Matrix.determinant
+import Mathlib.GroupTheory.Perm.Fin
+-- import Mathlib.Tactic.NormSwap
+import Qq
+
+open Qq
+
+-- todo: uncomment above imports when they are ported
+
+variables {α β : Type} [Semiring α] [Ring β]
+
+namespace Matrix
+
+open Matrix
+
+/-! Test that the dimensions are inferred correctly, even for empty matrices -/
+section dimensions
+
+-- set_option pp.universes true
+-- set_option pp.all true
+
+section elaborators
+open Lean Meta Elab Command
+
+/-- `dims% e` elaborates `e` as a Matrix and returns its dimensions as a `Nat × Nat`. -/
+elab "dims% " e:term : term => do
+  let elem_t ← mkFreshExprMVar (mkSort levelOne)
+  let m ← mkFreshExprMVar (mkConst ``Nat)
+  let n ← mkFreshExprMVar (mkConst ``Nat)
+  let matrix_t := mkAppN (mkConst ``Matrix [levelZero, levelZero, levelZero])
+                    #[mkApp (mkConst ``Fin) m, mkApp (mkConst ``Fin) n, elem_t]
+  let _ ← Term.elabTermEnsuringType e (some matrix_t)
+  let m ← instantiateMVars m
+  let n ← instantiateMVars n
+  mkAppM ``Prod.mk #[m, n]
+
+/-- `#guard e1 is e2` elaborates `e1` and `e2` and checks they are `BEq`. -/
```

Comment:
```suggestion
/-- `#guard e1 is e2` elaborates `e1` and `e2` and checks they are `BEq` as `Expr`s. -/
```

Reply:
I'll fix this. It'll be a `#guard_expr` command that more similarly matches the `guard_expr` tactic.

Reply:
Ok, it's now using the syntax from `guard_expr`, so it looks like
```
#guard_expr dims% !![1,2;3,4] =ₛ (2, 2)
```
It handles defeq testing too with `=~`:
```
#guard_expr dims% !![1,2;3,4] =~ (Nat.succ 1, 2)
```
Supposedly `guard_expr` handles `=` for equality at reducible transparency, but I can't get it to parse.

Reply:
> Supposedly `guard_expr` handles `=` for equality at reducible transparency, but I can't get it to parse.

You are probably forgetting to use `x:term:51 " = " y:term` or else `x` will also consume the `=` token.